### PR TITLE
crimson/osd: dissect PgOpsExecuter from OpsExecuter

### PR DESCRIPTION
**Pull request by rzarzynski**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/osd/ops_executer.h [hunk 1]
```diff
@@ -275,4 +265,23 @@ OpsExecuter::osd_op_errorator::future<> OpsExecuter::flush_changes(
   }
 }
 
+class PgOpsExecuter {
+  PG& pg;
```
Comment (tchaikov) [inline src/crimson/osd/ops_executer.h]: nit, might want to put the public parts earlier. see https://google.github.io/styleguide/cppguide.html#Declaration_Order

File: src/crimson/osd/ops_executer.h [hunk 2]
```diff
@@ -275,4 +265,23 @@ OpsExecuter::osd_op_errorator::future<> OpsExecuter::flush_changes(
...
+    return std::forward<Func>(f)(std::as_const(pg),
+                                 std::as_const(msg->get_hobj().nspace));
+  }
+
+public:
+  PgOpsExecuter(PG& pg, Ref<MOSDOp> msg)
+    : pg(pg), msg(std::move(msg)) {
+  }
+
+  seastar::future<> execute_pg_op(class OSDOp& osd_op);
```
Comment (tchaikov) [inline src/crimson/osd/ops_executer.h:277]:
```suggestion
  seastar::future<> execute_op(class OSDOp& osd_op);
```

as the name of the class implies that the ops being executed are PG Ops.

Comment (tchaikov) [reply src/crimson/osd/ops_executer.h:277]: for the same reason, might want to s/execute_osd_op/execute_op/ in `OpsExecuter`?

Comment (tchaikov) [reply src/crimson/osd/ops_executer.h:277]: also, might want to add a comment on top of `OpsExecuter` to explain that that class is solely for executing ops targeting objects. and add a comment on top of `PgOpsExecuter` to explain that that class is solely for executing ops targeting a certain PG.

File: src/crimson/osd/ops_executer.cc [hunk 3]
```diff
@@ -420,6 +420,148 @@ OpsExecuter::watch_errorator::future<> OpsExecuter::do_op_notify_ack(
   });
 }
 
+OpsExecuter::osd_op_errorator::future<>
```
Comment (tchaikov) [inline src/crimson/osd/ops_executer.cc:423]: > only move the code after PgOpsExecuter.

move `OpsExecuter::execute_osd_op()` closer to other methods of this class

File: src/crimson/osd/ops_executer.h (hunk 2 above)
Comment (rzarzynski) [reply src/crimson/osd/ops_executer.h:277]: > as the name of the class implies that the ops being executed are PG Ops.

Agreed, the duplication is a downside; the upside is easier grepping. No strong feelings here.

Comment (ronen-fr) [reply src/crimson/osd/ops_executer.h]: I'm for the easy grep. Analyzing current Ceph code is a slow process (without a good and fast IDE), in part due to this issue.

**Pull request merged**